File: Data Structures & Algorithms/combination-target-sum/submission-2.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        vector<int> current;
        backtrack(res, nums, current, 0, target, 0);
        return res;
    }

    void backtrack(vector<vector<int>>& res, vector<int>& nums, vector<int>& current, int sum, int target, int start) {
        if (sum > target) return;
        if (sum == target) {
            res.push_back(vector(current));
            return;
        }
        for (int i = start; i < nums.size(); i++) {
            current.push_back(nums[i]);
            backtrack(res, nums, current, sum + nums[i], target, i);
            current.pop_back();
        }
    }
};

File: Data Structures & Algorithms/subsets/submission-2.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        backtrack(ans, new ArrayList<>(), 0, nums);
        return ans;
    }

    public void backtrack(List<List<Integer>> ans, List<Integer> current, int n, int[] nums) {
        ans.add(new ArrayList<>(current));
        for (int i = n; i < nums.length; i++) {
            current.add(nums[i]);
            backtrack(ans, current, i + 1, nums);
            current.removeLast();
        }
    }
}
